examples/petsc/dmswarm.c

```diff
@@ -0,0 +1,410 @@
+//
+// Sample runs:
+//
+//     ./dmswarm
+
+/// @file
+/// libCEED example using PETSc with DMSwarm
+const char help[] = "libCEED example using PETSc with DMSwarm\n";
+
+#include <ceed.h>
+#include <petscdmplex.h>
+#include <petscdmswarm.h>
+#include <petscds.h>
+#include <petscfe.h>
+#include <petscksp.h>
+#include <petsc/private/petscfeimpl.h> /* For interpolation */
+
+#include "include/bpsproblemdata.h"
+#include "include/petscutils.h"
+
+static PetscScalar EvalU(PetscInt dim, const PetscScalar x[]) {
+  PetscScalar result = 1, center = 0.1;
+  for (PetscInt d = 0; d < dim; d++) {
+    result *= tanh(x[d] - center);
+    center += 0.1;
+  }
+  return result;
+}
+
+static PetscScalar EvaldU(PetscInt dim, const PetscScalar x[], PetscInt direction) {
+  PetscScalar result = 1, center = 0.1;
+  for (PetscInt d = 0; d < dim; d++) {
+    if (d == direction) result *= 1.0 / cosh(x[d] - center) / cosh(x[d] - center);
+    else result *= tanh(x[d] - center);
+    center += 0.1;
+  }
+  return result;
+}
+
+PetscErrorCode DMSwarmCreateReferenceCoordinates(DM dm_swarm, DM dm_mesh, IS *is_points, Vec *ref_coords) {
+  PetscInt           cell_start, cell_end, dim, num_points_local, *point_cell_numbers;
+  PetscScalar       *coords_points_ref;
+  const PetscScalar *coords_points_true;
+
+  PetscFunctionBeginUser;
+
+  // Create vector to hold reference coordinates
+  {
+    Vec true_coords;
+
+    PetscCall(DMSwarmCreateLocalVectorFromField(dm_swarm, "DMSwarmPIC_coor", &true_coords));
+    PetscCall(VecDuplicate(true_coords, ref_coords));
+    PetscCall(DMSwarmDestroyLocalVectorFromField(dm_swarm, "DMSwarmPIC_coor", &true_coords));
+  }
+
+  // Allocate index set array
+  PetscCall(VecGetLocalSize(*ref_coords, &num_points_local));
+  PetscCall(DMGetDimension(dm_mesh, &dim));
+  num_points_local /= dim;
+  PetscCall(PetscMalloc1(num_points_local, &point_cell_numbers));
+
+  // Get reference coordinates for each swarm point wrt the elements in the background mesh
+  PetscCall(DMSwarmSortGetAccess(dm_swarm));
+  PetscCall(DMPlexGetHeightStratum(dm_mesh, 0, &cell_start, &cell_end));
+  PetscCall(DMSwarmGetField(dm_swarm, "DMSwarmPIC_coor", NULL, NULL, (void **)&coords_points_true));
+  PetscCall(VecGetArray(*ref_coords, &coords_points_ref));
+  for (PetscInt cell = cell_start; cell < cell_end; cell++) {
+    PetscReal *coords_cell_points_true, *coords_cell_points_ref;
+    PetscInt  *cell_points;
+    PetscInt   num_points_in_cell;
+
+    PetscCall(DMSwarmSortGetPointsPerCell(dm_swarm, cell, &num_points_in_cell, &cell_points));
+    PetscCall(DMGetWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_cell_points_true));
+    PetscCall(DMGetWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_cell_points_ref));
+    // -- Reference coordinates for swarm points in background mesh element
+    for (PetscInt p = 0; p < num_points_in_cell; p++) {
+      point_cell_numbers[cell_points[p]] = cell;
```

Comment:
Maybe this? I'm still deciding exactly what to put in this index set.
```suggestion
      point_cell_numbers[cell_points[p]] = cell - cell_start;
```

Related - we should decide what sort of restriction operation we need to create for this and how to update it. Of note for this restriction is the fact that a) we have different numbers of swarm points per cell 2) we need to update the mapping or create/destroy this restriction once per time step

Reply:
For the particle restriction, two options come to mind:

1. Sort the particles by cell and provide the offsets `CeedInt particle_offset[num_cells + 1]`. These are similar to row starts in a CSR matrix format. The particles for each cell are contiguous in memory.
2. Create an integer permutation array `particle_index[num_particles]` so that the particles themselves don't need to be sorted. This requires less up-front work, but then a less regular memory access pattern.

I think either is fine, though I don't know if `DMSwarm` supports that intrusive sort (which has to also reorder every field). The second option is more flexible, especially if different operators need to work over different parts of the domain (so that the particles they need to work cannot be contiguous for each operator). That makes me lean toward 2.

Reply:
As we dig more into this - I think what libCEED really needs is all of the indices of the points in the first local cell, followed by all the indices of the points in the second local cell, etc
something like
4
6
84
102
-1
23
35
76
-1
-1
46
78
...

where -1 indicates a break between cells

right now we have a nested loop with num_elem * num_local_points iterations all over the place, but this would bring us down to num_elem * ~num_points_per_elem, which is much, much fewer iterations

Reply:
resolved per conversation in Zulip - see changes in latest commit